### ISSUE TITLE
Code printers can't print Sum

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#10194]
```
In [2]: ccode(Sum(k, (k, 1, n)))
Out[2]: '// Not supported in C:\n// tuple\nSum(k, (k, 1, n))'
```

And similar for `fcode`.

https://github.com/sympy/sympy/pull/9314 may be relevant here.

**Comment by khannavansh** [comment]
Can you elaborate on the issue a bit more , I am interested in handling this bug .

**Comment by asmeurer** [comment]
`ccode(Sum(k, (k, 1, n)))` should print valid `C` code that computes the summation, i.e., a loop. Ideally it should be something that works inline.

**Comment by khannavansh** [comment]
@asmeurer  Thank you , got it.

**Comment by asmeurer** [comment]
Any ideas how to do this? `ccode` should generate an expression, but I don't know if a summation loop can be written as a single expression.

I checked Mathematica's CForm (side note: you can run Mathematica for free in the cloud at https://lab.wolframcloud.com/), but it just uses a `Sum` function which is defined in `mdefs.h`, which you have to include in your code. Same for FortranForm.

**Comment by moorepants** [comment]
We've talked about including our own set of functions for the supported languages. There are some other issues where this would be appropriate.

**Comment by asmeurer** [comment]
What are some of the other places where this is needed?

**Comment by moorepants** [comment]
I think it came up in the discussion about printing Pow. I know @jcrist has talked about it when he was doing his GSoC work. I just searched our issues and didn't find anything, but I know there are some conversations about us providing a mini C header with various functions.

**Comment by jcrist** [comment]
The reason for including our own header would be to implement things like `powi`, for faster integer exponentiation, and also special functions (although a PR to scipy to include the cephes headers on install would probably be better for that). 

Back when I was looking at this I started experimenting with rewriting the codegen stuff using a more ast based formation. This would allow for easier composable generation of multi-line expressions (loops, imports/includes, declarations, headers, etc...). The code is here: https://github.com/jcrist/symcc. I know everything in there currently *works*, but I got side tracked with school and never finished it. I still think this is a better design in the end, as expressions like `Sum` can be rewritten to instances of `For`, and then generated as a loop by the codegenerator. Separation of concerns and all that.

**Comment by asmeurer** [comment]
Thanks for sharing. I've been thinking about doing the same things (see https://github.com/sympy/sympy/wiki/Code-Generation-Notes). I'll definitely take a closer look at symcc. I already started playing with my own CodeBlock type for blocks of assignments, but then I got sidetracked with https://github.com/sympy/sympy/pull/10271.

A few thoughts/concerns:

- I don't want to get in the business of writing our own numerical routines in C (or Fortran or whatever). Small functions like sum or powi are probably fine. For special functions, we should just require an external header. 

- I'm trying to think about the modularity of the codegen module(s). Right now, `ccode` returns a single expression (except it can also return a statement if you use `assign_to`, but let's ignore that). Requiring a header file breaks that a little bit because there's no way for ccode to indicate that the header is required (other than in its docstring), since the return type is an expression string. Ideally, one layer up (what is now the `codegen` layer, but will probably end up being replaced with something better, like your symcc) should be what knows about details like function definitions, header files, and so on (anything external to the statement itself). 

I suppose this is a leaky abstraction. There's no way for `ccode`, etc. in general to produce completely standalone expressions in every case.

**Comment by moorepants** [comment]
This is an aside, but I've been reading about this project over the weekend: https://github.com/antlr/antlr4. Might have some good ideas and also may show that we should use something like this as a dependency. Note that someone quickly wrote a latex2sympy parser with it: https://github.com/augustt198/latex2sympy

**Comment by asmeurer** [comment]
I discussed this a bit with @scopatz the other day. I think it's best for the code printers to only print expressions. There is already some precedence for this:

```
In [13]: fcode(Piecewise((x, Eq(x, 1)), (y, Eq(x, 2)), (z, True)), standard=77)
...
NotImplementedError: Using Piecewise as an expression using inline operators is not supported in standards earlier than Fortran95.
```

(I also think that regardless of this stance that ccode should be raising an exception, not returning a string, on error)

This means that you can always take the output of `ccode`, `fcode`, etc. and interpolate it in some larger expression, or assign it directly to a variable or whatever. 

There needs to be a higher level abstraction (I like @jcrist's symcc) for printing statements, blocks of code, function definitions, and entire modules. 

There would then be a function or set of functions for this level for printing things. Most people would be encouraged to use this level, unless they explicitly want expressions only, especially since the most popular languages (C and Fortran) are limited on what sorts of things can go in an expression (to contrast, it is trivial to perform a sum in an expression in Python, using the `sum` builtin, and even without it, you can do virtually anything in a single expression by sufficient use of lambdas and things like list comprehensions). 

Right now this level of abstraction is the codegen module, but I'm not a fan of it. I've not really [seen anyone use it](https://github.com/sympy/sympy/wiki/Code-Generation-Notes) (outside of SymPy itself in autowrap), and I have seen a lot of people write their own thing to do what it is trying to do. 

We also discussed the idea of having a header file, and he was opposed to it (and I think I agree). It's better to write code that is self contained. This makes compilation much easier, as the user doesn't have to worry about linking in some library. It also makes it more user-friendly, as someone can just use `ccode` (for example) and get a workable expression, or an exception pointing them to another function that would do the same, whereas with a required header they might not know that they needed it (if they didn't read the documentation carefully), and be confused about why such-and-such function is missing. 

I think not having a header just means inlining the definitions for things (like sum). But this is fine, because generated code need not be compact or even readable, and it opens the possibility for more compiler optimizations. The only real downside, from what I can see, is that functions like `ccode` that should return expressions only will be limited in what they can print (like probably not Sum, unless I am missing some way to do it), but if they can instead raise an error pointing to some other function that can, I think that is fine. 

@scopatz please feel free to add to this if I left something out or misremembered your opinions.

**Comment by moorepants** [comment]
FYI, this search shows 467 files on Github with "sympy.utilities.codegen" in them:

https://github.com/search?utf8=%E2%9C%93&q=%22sympy.utilities.codegen%22+in%3Afile+filename%3A*.py&type=Code&ref=searchresults

**Comment by asmeurer** [comment]
Cool. So some people do indeed use it. I still stand by what I said. I personally found it rather clunky. It seems like a good tool if it does exactly what you are looking for (spit out one module with one function that computes your expression), but it doesn't seem to be very easy to customize or extend (the Routine objects in the code are particularly confusing, at least to me), and for even slightly different use-cases (like including code as part of a larger module), it doesn't work at all, which I believe is one reason why I see people doing so much templating stuff (templating is not inherently bad, but it's a slippery slope from doing templating to doing "+= code generation" and "stringy calculations" which are difficult to read and maintain and lose the advantages of doing things symbolically in the first place).

**Comment by asmeurer** [comment]
This is all a bit off topic. For this issue (code printing Sum), I suggest merging @jcrist's symcc into SymPy and implementing it in that. Unless someone knows of a way to do a summation as an expression in C.

**Comment by moorepants** [comment]
I agree, let's not implement things that can't be expressed as one line. We can discuss the bigger code gen issues in other issue or on the mailing list.

**Comment by cbm755** [comment]
(+1 for difficulty in understanding `Routine`: I had a hard time with that when adding Octave support a year ago.)

Ironically, it worse for `octave_code` where we actually could generate:
````
>>> octave_code(Sum(k**2, (k, 1, n)))
sum([1:n].^2)
````
But instead we appear to just call `str` (!)

**Comment by jiboncom** [comment]
I understand why you would like to avoid returning code blocks in `ccode`. But, even the Sum loop could be done in one line, how could you declare the summands? Are there anonymous functions in C? I thought there were not.

**Comment by jiboncom** [comment]
I got lost on the many issues about this but somewhere @asmeurer mentioned he had discussed using code blocks in C (statements enclosed in curly brackets). I don't know what came out of that discussion but I managed to use them to implement this and even get them to work with autowrap. It also works with nested sums!

I made a Colab if you want to check what I did. Let me know what you thing and if it looks good, I will be happy to make a PR with it: https://colab.research.google.com/drive/1LyYVhCMRxJ-W3EPtpErapbzKTEhfGaqY?usp=sharing

The only issue is that I am having a few problems with autowrap on Windows 11 but it works fine in Linux. Any ideas?

**Comment by ThePauliPrinciple** [comment]
It seems like a good approach. I don't think it works for non-scalars though:

```python
from sympy import MatrixSymbol, Sum, Idx
i, j = Idx('i', 5), Idx('j', 5)
expr = Sum(MatrixSymbol('A', 5,5)[i,j], i)
myPrinter().doprint(expr)
```
Not sure what it should do in that case (similarly for `ArraySymbol` or `Indexed`)

**Comment by ThePauliPrinciple** [comment]
There is a `Dummy` object in sympy which handles being printed differently for every instance of it, you can use that instead of the random number that's used now.

**Comment by jiboncom** [comment]
Yes, I like the idea of using a `Dummy` to hold the results. About your non-scalar example, I am not sure either. As I see it, you're *summing the elements of the j th column of A* with j being a known parameter or user defined input (when using autowrap). That would be easy add when the second argument of the sum (i) is a ranged `Idx`. Does Sympy do something different? Would you expect something different?

**Comment by ThePauliPrinciple** [comment]
The printer has a setting called `contract`, it should automatically loop over indices if that is set to true.
```python
from sympy import Eq, IndexedBase, Idx, Sum
from sympy.printing.c import ccode
len_y = 5
y = IndexedBase('y', shape=(len_y,))
z = IndexedBase('y', shape=(len_y,))
x = IndexedBase('x', shape=(len_y,len_y))
i = Idx('i', len_y-1)
j = Idx('j', len_y-1)
#this will print loops for us
print(myPrinter().doprint(Assignment(y[i],z[i])))
#this fails
print(myPrinter().doprint(Assignment(y[i],Sum(x[i,j], j))))
```
Because `get_indices` does not handle the case of `Sum`. I would assume that it should remove the index (or indices) that is already summed over from the set of indices returned by that function.

**Comment by asmeurer** [comment]
@ThePauliPrinciple is right that the real testbed for this sort of thing is printing matrix expression constructs. The original motivator for this issue was printing a dot product. 

The idea with CodeBlock, and using the sympy.codegen.ast classes generally, is that the generated code here can get complicated quickly. So it's better to construct the loop and so on directly as an ast, then print that, rather than trying to manage all this on the strings in the printer. It also means that the same thing can be reused for many languages, and each language only need to know how to print the basic constructs like `For`. Even for simpler constructs, manipulating asts will be more maintainable than manipulating strings in the printers.

It also opens up a lot of options for optimizing things in the codegen ast before printing it (for example, the "assigning the intermediate result to 0" issue mentioned by @ThePauliPrinciple could be fixed in the ast before printing). Of course many optimizations can already be done by compilers, so we should focus on actually implementing optimizations that are harder for compilers. 

The other reason to use codegen asts is that it allows doing nonlocal things, or things that are impossible as expressions. C is very lenient in what it will allow as an expression but other languages aren't. Even in C, the syntax `for (int i = 0; ...)` is C99+ only (https://stackoverflow.com/questions/1287863/c-for-loop-int-initial-declaration). So in general we would need to keep track of the initialization of the new variable to insert it into the declarations somewhere else. I think also with C you might need to worry about the way scoping works in blocks.

This is the long-term goal for codegen.

**Comment by jiboncom** [comment]
I am not exactly sure what you mean with that solution not working for printing a dot product, because I think this would work. But I do get the advantages of AST and the importance of providing compatibility to more languages.

I made a working POC of how that could work. My idea was:

- Look for subexpressions that need a Codeblock* implementation.
- Replace all those subexpressions with a result variable.
- Have the `self._print_*` function generate the code block to resolve the value of the result variable.
- Alter the `doprint` function to first print the code blocks and then the expression with the substitutions
- Be able to do this recursively

It is easier to share this Colab with the code rather than pasting different examples here. 
https://colab.research.google.com/drive/1U6inffXS32VWb6sW4AfkEFVyhLAm3ApK#scrollTo=Zq2-_Q0Fgr9J

* for now is only `sympy.Sum` but ideally it could be a new inheritable class to check for. Then this could work with any function that needs a Code Block to be generated!

**Comment by asmeurer** [comment]
The way I think about it is that you have multiple stages. The ast nodes are like an IR in a compiler. They would get generated first, and then the printer would get called on the ast nodes instead of the expression itself. 

Separating the node generation from the printing allows doing things with the codegen nodes before printing them, e.g., applying special optimization or rewrite passes to them (for example, `CodeBlock.cse`).

**Comment by jiboncom** [comment]
Not sure if this is exactly what you were thinking but it is simple and works well without breaking any existing pipeline. 

`doprint(expr)` works just as it does now. But it would be nice if it raised a `Warning` if the expression contains a function for which there is a `_codeblock_TypeName()` method defined in the printer class. This warning suggests calling a new `doblocks(expr)` method before printing the expression.

The `doblocks(expr)` method returns `expr` with all the highest level `CodeBlock Functions` replaced with a new `codeblockResult` object that inherits from `Dummy` and the respective `CodeBlock` object that assigns value to each `codeblockResult`. Each `_codeblock_TypeName()` method calls `doblocks(expr)` on its arguments and deals with the possible nested coded blocks on its own. This gives a `new_expr` that can be used as a one-liner and all the `CodeBlock`s that need to be run before using `new_expr` anywhere.

Finally, code generation would be made as follows:
```python
# Code noting function is not implemented and suggestion to use doblocks() first
myPrinterObj.doprint(expr) 

new_expr, code_blocks = myPrinterObj.doblocks(expr)

for code_block in code_blocks:
  print(myPrinterObj.doprint(code_block))
  
print(myPrinterObj.doprint(new_expr))
```

I made a quick working demo here: 
https://colab.research.google.com/drive/18y0Vat9wtrj2OXen6DsWGLujOURA3V9h

PS: There is a bug when printing `CodeBlock`s in the C printers that adds a semicolon to AST `Comment`s and to empty `CodeBlock`s

**Comment by asmeurer** [comment]
This seems like a reasonable approach. I'd have to think a bit whether there's any potential issues with it. 

The main way I like to think about this is the distinction between expressions and statements (or blocks of statements), which most programming languages use. The current printers are based around printing expressions, which makes things simple because expressions can always be recursively nested within each other. But some constructs like loops and variable assignments can't be represented as expressions in some languages. Things are a little complicated because different languages have different rules about expressions vs. non-expressions and ideally we'd use a design that is language independent. So one thing to think about is whether your design will work for the other languages we support (e.g., Python, Fortran, C++, + maybe also some languages with more esoteric syntax rules).

**Comment by asmeurer** [comment]
Yes, please do. It will be much easier to discuss the specifics of the approach if it is in a PR. There are many others who will likely have opinions as well, including everyone else who has commented on this issue plus others like @bjodah.

**Comment by jiboncom** [comment]
Ok, done! Let's continue the discussion there.
